ksql-engine/src/main/java/io/confluent/ksql/planner/plan/DataSourceNode.java

```diff
@@ -158,294 +127,35 @@ public int getPartitions(final KafkaTopicClient kafkaTopicClient) {
 
   @Override
   public SchemaKStream<?> buildStream(final KsqlQueryBuilder builder) {
-    final Optional<WindowInfo> windowInfo = dataSource
-        .getKsqlTopic()
-        .getKeyFormat()
-        .getWindowInfo();
-
-    if (windowInfo.isPresent()) {
-      final KeySerdeSupplier<Windowed<Struct>> keySerdeSupplier =
-          (format, physicalSchema, logContext) ->
-              builder.buildKeySerde(format, windowInfo.get(), physicalSchema, logContext);
-
-      return buildStream(builder, keySerdeSupplier, windowedMapper());
-    }
-
-    return buildStream(builder, builder::buildKeySerde, nonWindowedValueMapper());
-  }
-
-  private <K> SchemaKStream<K> buildStream(
-      final KsqlQueryBuilder builder,
-      final KeySerdeSupplier<K> keySerdeSupplier,
-      final ValueMapperWithKey<K, GenericRow, GenericRow> mapper
-  ) {
-    final Stacker contextStacker = builder.buildNodeContext(getId());
-
-    final Consumed<K, GenericRow> consumed =
-        buildSourceConsumed(builder, keySerdeSupplier, contextStacker);
-
-    return getDataSourceType() == DataSourceType.KTABLE
-        ? buildSourceTable(builder, consumed, mapper, contextStacker)
-        : buildSourceStream(builder, consumed, mapper, contextStacker);
-  }
-
-  private <K> SchemaKTable<K> buildSourceTable(
-      final KsqlQueryBuilder builder,
-      final Consumed<K, GenericRow> consumed,
-      final ValueMapperWithKey<K, GenericRow, GenericRow> mapper,
-      final Stacker contextStacker
-  ) {
-    final KTable<K, GenericRow> kTable = createKTable(builder, consumed, mapper, contextStacker);
-
-    final KeySerde<K> keySerde = (KeySerde<K>) KsqlConsumed.keySerde(consumed);
-
-    return new SchemaKTable<>(
-        kTable,
-        schema,
-        keySerde,
-        keyField,
-        ImmutableList.of(),
-        SchemaKStream.Type.SOURCE,
-        builder.getKsqlConfig(),
-        builder.getFunctionRegistry(),
-        contextStacker.getQueryContext()
-    );
-  }
-
-  private <K> SchemaKStream<K> buildSourceStream(
-      final KsqlQueryBuilder builder,
-      final Consumed<K, GenericRow> consumed,
-      final ValueMapperWithKey<K, GenericRow, GenericRow> mapper,
-      final Stacker contextStacker
-  ) {
-    final KStream<K, GenericRow> kstream = createKStream(builder, consumed, mapper);
-
-    final KeySerde<K> keySerde = (KeySerde<K>) KsqlConsumed.keySerde(consumed);
-
-    return new SchemaKStream<>(
-        kstream,
+    final Stacker contextStacker = builder.buildNodeContext(getId().toString());
+    final SchemaKStream<?> schemaKStream = schemaKStreamFactory.create(
+        builder,
+        dataSource,
         schema,
-        keySerde,
-        keyField,
-        ImmutableList.of(),
-        SchemaKStream.Type.SOURCE,
-        builder.getKsqlConfig(),
-        builder.getFunctionRegistry(),
-        contextStacker.getQueryContext()
+        contextStacker.push(SOURCE_OP_NAME).getQueryContext(),
+        builder.getAutoOffsetReset(),
```

Comment:
if we're passing in the builder, we shouldn't need to pass this in independently